src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -1002,54 +1002,77 @@ RandomBlockOolWriter::do_write(
          t, extents.size());
   std::vector<write_info_t> writes;
   writes.reserve(extents.size());
+
   for (auto& ex : extents) {
-    auto paddr = ex->get_paddr();
-    assert(paddr.is_absolute());
-    RandomBlockManager * rbm = rb_cleaner->get_rbm(paddr); 
-    assert(rbm);
-    TRACE("write extent {}, paddr {} ...",
-          fmt::ptr(ex.get()), paddr);
     auto& stats = t.get_ool_write_stats();
     stats.extents.num += 1;
     stats.extents.bytes += ex->get_length();
-    stats.num_records += 1;
-
     ex->prepare_write();
-    extent_len_t offset = 0;
+
+    RandomBlockManager * rbm = rb_cleaner->get_rbm(ex->get_paddr()); 
+    assert(rbm);
+    paddr_t cur_base = ex->get_paddr();
+    assert(ex_base.is_absolute());
+    extent_len_t cur_len = ex->get_length();
     bufferptr bp;
+    TRACE("write extent {}, paddr {} ...",
+	  fmt::ptr(ex.get()), cur_base);
     if (can_inplace_rewrite(t, ex)) {
       assert(ex->is_logical());
       auto r = ex->template cast<LogicalCachedExtent>()->get_modified_region();
       ceph_assert(r.has_value());
-      offset = p2align(r->offset, rbm->get_block_size());
-      extent_len_t len =
-	p2roundup(r->offset + r->len, rbm->get_block_size()) - offset;
-      bp = ceph::bufferptr(ex->get_bptr(), offset, len);
+      extent_len_t offset = p2align(r->offset, rbm->get_block_size());
+      cur_len = p2roundup(r->offset + r->len, rbm->get_block_size()) - offset;
+      cur_base = ex->get_paddr() + offset;
+      bp = ceph::bufferptr(ex->get_bptr(), offset, cur_len);
     } else {
       bp = ex->get_bptr();
       auto& trans_stats = get_by_src(w_stats.stats_by_src, t.get_src());
-      ++(trans_stats.num_records);
       trans_stats.data_bytes += ex->get_length();
       w_stats.data_bytes += ex->get_length();
     }
-    writes.push_back(write_info_t{paddr + offset, std::move(bp), rbm});
 
     if (ex->is_initial_pending()) {
       t.mark_allocated_extent_ool(ex);
     } else if (can_inplace_rewrite(t, ex)) {
       assert(ex->is_logical());
       t.mark_inplace_rewrite_extent_ool(
-        ex->template cast<LogicalCachedExtent>());
+	ex->template cast<LogicalCachedExtent>());
     } else {
       ceph_assert("impossible");
     }
+
+    // TODO : allocate a consecutive address based on a transaction
+    if (writes.size() != 0 &&
+       (writes.back().offset + writes.back().len == cur_base &&
+       writes.back().bp.c_str() + writes.back().len == ex->get_bptr().c_str())) {
```

Comment:
Not following `writes.back().bp.c_str() + writes.back().len == ex->get_bptr().c_str()` ?

Reply:
This checks if the previous `bp`'s address and the current extent's `bp`'s address are aligned in a row. As a result,  when submitting ` rbm->write`, we only need a bp without performing a memory copy. If we allocate consecutive memory blocks based on a transaction, the likelihood of this merging will increase.

Reply:
The precondition of this optimization seems too strict:
1. The two extents must be consecutive in `extents`
2. Their memory addresses of the buffers should be consecutive
3. Their physical addresses should be consecutive 

If I understand the "TODO" correctly, seems to me 2 is unlikely until "TODO" is working correctly. How about using memory copy to replace 2 for now?

Reply:
Will do.